src/scheduler/MpiWorld.cpp

```diff
@@ -912,53 +918,138 @@ void MpiWorld::gather(int sendRank,
 
     bool isInPlace = sendBuffer == recvBuffer;
 
-    // If we're the root, do the gathering
     if (sendRank == recvRank) {
         SPDLOG_TRACE("MPI - gather all -> {}", recvRank);
 
-        // Iterate through each rank
-        for (int r = 0; r < size; r++) {
-            // Work out where in the receive buffer this rank's data goes
-            uint8_t* recvChunk = recvBuffer + (r * recvOffset);
-
-            if ((r == recvRank) && isInPlace) {
-                // If operating in-place, data for the root rank is already in
-                // position
-                continue;
-            } else if (r == recvRank) {
-                // Copy data locally on root
-                std::copy(sendBuffer, sendBuffer + sendOffset, recvChunk);
+        for (auto it : ranksForHost) {
+            if (it.first == thisHost) {
+                // Each local rank will send a separate message that we receive
+                // directly in the receive buffer.
+                for (const int r : it.second) {
+                    if (r == recvRank) {
+                        // If not receiving in place, copy our data to the
+                        // right offset
+                        if (!isInPlace) {
+                            memcpy(recvBuffer + (recvRank * recvOffset),
+                                   sendBuffer,
+                                   sendOffset);
+                        }
+
+                        continue;
+                    }
+
+                    recv(r,
+                         recvRank,
+                         recvBuffer + (r * recvOffset),
+                         recvType,
+                         recvCount,
+                         nullptr,
+                         faabric::MPIMessage::GATHER);
+                }
             } else {
-                // Receive data from rank if it's not the root
-                recv(r,
+                // Each remote leader will send a signle message with the data
+                // for all its local ranks
+                auto rankData =
+                  std::make_unique<uint8_t[]>(it.second.size() * recvOffset);
+
+                recv(it.second.front(),
                      recvRank,
-                     recvChunk,
+                     rankData.get(),
                      recvType,
-                     recvCount,
+                     recvCount * it.second.size(),
                      nullptr,
                      faabric::MPIMessage::GATHER);
+
+                // Copy each received chunk to its offset
+                for (int r = 0; r < it.second.size(); r++) {
```

Comment:
This is necessary as we can not assume that ranks for the same host are always consecutive numbers.

The way MPI works atm this only happens if we are overloading the master after running out of resources.

Reply:
We construct this vector on world creation time don't we? Can we sort it at that point to simplify this code by knowing that ranks are consecutive?

Reply:
Yes. What I was trying to say here though is that if rank X, local leader of host Y, is sending the contribution of all its local ranks including itself `[X, X1, X2, X3]`, we can't assume `X, X1, X2, X3` to be consecutive numbers.

As a consequence, the receiver of the gather (that has alocated a buffer with `worldSize` slots), can't just copy the four coalesced bits of information starting at `X`s offset.